### DISCUSSION TITLE
feature for using of self created and calibrated map ( as possible in/with ublox )

### DISCUSSION THREAD

**Discussion by wernerdd** [semuconsulting/PyGPSClient#115] [Ideas and Feature Requests]
It would be fine to have a similar feature as with the u-blox-tool  ( [ubx-User-Guide](https://content.u-blox.com/sites/default/files/u-center_Userguide_UBX-13005250.pdf) , chapter 5.2.10.2 Map calibration ).
So you can use the pygpsclient offline or if working with NTRIP one has just low data transfer and no key for online-pas is needed.
So one can use your own scanned map as possible for other GPS-Devises too ( e.g. Garmin, ... look [ExperetGPS](https://www.expertgps.com/help/calibrating-a-scanned-map.asp) ).

**Comment by semudev2** [comment]
Hi @wernerdd 

Thanks for the suggestion. Probably wouldn't be that hard to implement - I effectively already do something similar for the World Map display (see https://github.com/semuconsulting/PyGPSClient/blob/master/src/pygpsclient/map_frame.py#L173) and could readily adapt this to larger scale map tiles.

For simplicity sake I'd probably want to add a few provisional caveats:

1. User-supplied map must be a standard WGS84 Web Mercator projection (i.e. lon and lat lines are parallel).
2. User-supplied map must be in png format (user can use an external tool to convert formats).
3. User would provide file path and calibration points via a new 'user map' section in the pygpsclient.json configuration file.

Happy to add it to the feature backlog for implementation as and when I get the time.

**Comment by semuadmin** [comment]
Hi @wernerdd,

I've added an enhancement which implements basic support for a user-defined custom (offline) map:

The user-defined map can be defined via the following configuration file parameters:

```
"usermaps_l": [
   [
      "/home/myuser/maps/mymap01.tif",
      [52.48903090029343, 13.376796709434993, 52.428992219505325, 13.513273402895]
   ],
   [
      "/home/myuser/maps/mymap02.tif",
      [52.43227595237385, 13.499564532600285, 52.37215985772205, 13.636041225870589]
   ],
   [
      "/home/myuser/maps/mymap03.tif",
      [52.37776419675291, 13.625147570124257, 52.31757380344052, 13.761624263584265]
   ],
   etc...
]
```

If the current receiver position is outside the bounds of the map, you'll see an 'OUT OF BOUNDS' warning message.

See the [sample pygpsclient.json configuration file](https://github.com/semuconsulting/PyGPSClient/blob/master/pygpsclient.json). Here's an illustration using a custom geoTIFF image exported from an [Open Street Map](https://www.openstreetmap.org/#map=10/52.4807/13.3443) © tile using [QGIS](https://qgis.org/en/site/):

<img width="478" alt="Screenshot 2024-02-21 at 15 35 17" src="https://github.com/semuconsulting/PyGPSClient/assets/28569967/f205e538-3970-4db0-a85f-ca78b7c21925">

**Comment by gdt** [comment]
It would be nice to support geotiff so you can export a map in qgis and use it, if there is custom map support at all.

**Reply by semuadmin** [reply]
Could readily support any image file format that can be opened as an [`PIL.ImageTk.PhotoImage`](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html) and loaded into a tkinter canvas (including PNG, JPG, GIF, TIFF, WEBP, etc. - the list is dependent on underlying operating system support), but would need some extra work to convert CRS and generate lat/lon bounding box (e.g. using `rasterio` library).

In the short term you could use something like this to automatically generate the necessary PyGPSClient json config settings for your geoTIFF file:

```python

from rasterio import open as openraster
from rasterio.warp import transform_bounds

files = ["mymap01.tif", "mymap02.tif", "mymap03.tif"]

print('"usermaps_l": [')
for i, fl in enumerate(files):
   fl = fl.strip()
   ras = openraster(fl)
   lonmin, latmin, lonmax, latmax = transform_bounds(ras.crs.to_epsg(), 4326, *ras.bounds)
   print( f'   [\n      "{abspath(fl)}",\n      {[latmax, lonmin, latmin, lonmax]}\n   ]{"," if i < len(files)-1 else ""}')
print("]")
```

```shell
"usermaps_l": [
   [
      "/home/myuser/maps/mymap01.tif",
      [52.48903090029343, 13.376796709434993, 52.428992219505325, 13.513273402895]
   ],
   [
      "/home/myuser/maps/mymap02.tif",
      [52.43227595237385, 13.499564532600285, 52.37215985772205, 13.636041225870589]
   ],
   [
      "/home/myuser/maps/mymap03.tif",
      [52.37776419675291, 13.625147570124257, 52.31757380344052, 13.761624263584265]
   ]
]
```

**Comment by semuadmin** [comment]
Basic offline custom map support has been added in pygpsclient>=1.4.11, as described above.

I've added a simple utility [georef.py](https://github.com/semuconsulting/PyGPSClient/blob/master/examples/georef.py) in the \examples folder which can automatically generate the relevant PyGPSClient configuration settings for any supported georeferenced image files e.g. geoTIFF.

I may make automatic geoTIFF import a core function in a future release, but for the time I am keeping this import utility separate as I would prefer to avoid an additional external library dependency (`rasterio`) for functionality which may only be used infrequently.

Feedback and contributions welcome.
